### PR TITLE
Add instructional error message for cudnn RNN double backward workaround

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33884 Add instructional error message for cudnn RNN double backward workaround**

Mitigates https://github.com/pytorch/pytorch/issues/5261.

It's not possible for us to support cudnn RNN double backwards due to
limitations in the cudnn API. This PR makes it so that we raise an error
message if users try to get the double backward on a cudnn RNN; in the
error message we suggest using the non-cudnn RNN.

Test Plan:
- added some tests to check the error message

Differential Revision: [D20143544](https://our.internmc.facebook.com/intern/diff/D20143544)

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -56,9 +59,21 @@ void copy_range(variable_list& out, IndexRange range, at::ArrayRef<Tensor> t) {
   std::copy(t.begin(), t.end(), out.begin() + range.first);
 }
 
-Tensor not_implemented(const char* name) {
-  throw std::runtime_error(
-      std::string("the derivative for '") + name + "' is not implemented");
+template <typename T>
+T not_implemented_base(const char* name, const char* reason) {
+  std::string msg = std::string("the derivative for '") + name + "' is not implemented.";
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: nit: I usually do this with `c10::str("the derivative for '", name, "' is not implemented.");` as it looks slightly better.

**Review by albanD — DISMISSED** [review]
Only minor nit, thanks!

**Comment by zou3519** [comment]
test_cross_device_reentrant_autograd_cuda seems to consistently fail on the CUDA job for me, but I can't reproduce it locally or while ssh-ing into a CircleCI machine. This will need more investigation.

**Comment by zou3519** [comment]
Edit: I figured out how to repro the error locally. It required running two tests in a specific order. https://gist.github.com/zou3519/17aa029509c83470aa299666ceab1d6f.

It looks like what happening is that raising the double backward error message permanently messes with the "checkpoint state" of the engine. Will dig into this a bit more later.

**Comment by albanD** [comment]
@zou3519 yes there is a known issue that is state is not reset properly: https://github.com/pytorch/pytorch/issues/37874
Is the test that leave the engine in a bad state the one you added? You might want to modify it to avoid doing so (at least until that other issue is fixed).

**Comment by zou3519** [comment]
> #37874

Ah, that makes sense, thanks for pointing it out! Yes, the tests I added check for the existence of the error message... we can remove the tests for now as a workaround (leaving this PR untested, lol)

**Comment by albanD** [comment]
Yes this bug is pretty nasty for testing...

I think you can work around this by running the double backward that fails by hand with a `.backward()` that way it will still raise but will also leave the state clean.

**Comment by zou3519** [comment]
> Yes this bug is pretty nasty for testing...
> 
> I think you can work around this by running the double backward that fails by hand with a `.backward()` that way it will still raise but will also leave the state clean.

Thanks! That worked out great

**Review by albanD — APPROVED** [review]
Thanks for the updated test!

**Pull request closed without merging**